mmengine/hooks/early_stopping_hook.py

```diff
@@ -0,0 +1,98 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import warnings
+from typing import List, Optional, Union
+
+from mmengine.registry import HOOKS
+from .hook import Hook
+
+DATA_BATCH = Optional[Union[dict, tuple, list]]
+
+
+@HOOKS.register_module()
+class EarlyStoppingHook(Hook):
+    """Early stop the training when the metric reached a plateau.
+
+    Args:
+        metric (str): The metric key to decide early stopping.
+        rule (str, optional): Comparison rule. Options are 'greater',
+            'less'. Defaults to None.
+        delta(float, optional): Minimum difference to continue the training.
+            Defaults to 0.01.
+        pool_size (int, optional): The number of experiments to consider.
+            Defaults to 5.
+        patience (int, optional): Maximum number of tolerance.
+            Defaults to 0.
```

Comment:
Hi, referring to https://github.com/Bjarten/early-stopping-pytorch/blob/master/pytorchtools.py
Training will be stopped when the metric gets consistently worse "patience" times. The `rule` argument means the better trends we want, and the patience means How long to wait after last time validation metric improved. Such setting makes sense to me.

If I understand correctly, the current implementation will stop training when the metric gets consistently better "patience" times, is it our expected?

Reply:
BTW, I think we could also adopt the strategy here https://github.com/open-mmlab/mmengine/pull/739#pullrequestreview-1185414869

If we reach the target metric threshold, just stop training.

Reply:
Inspired by [`ray.tune.stopper`](https://docs.ray.io/en/latest/_modules/ray/tune/stopper/experiment_plateau.html#ExperimentPlateauStopper),
my original intention was "Stop training when the current score does not exceed "the delta plus average of top `pool_size` scores. But it is acceptable to `patience` times when the conditions are not met.".

In this aspect, `rule` means that "current score should be `'greater|less'` than the average of latest scores.".
So, to check when the condition does not meet, we put the 'the average of latest scores' as a left argument.
```python
compare(sum(self.pool_values) / self.pool_size + self.delta, latest_value)
```

To my understanding, the current implementation meets your description. Is it better to improve the clarity?

Reply:
Ah, I don't know what happened, I misunderstood in the morning :laughing: ,my bad :rofl: 

But maybe we should set the default value of patience to 1 or a larger one? I think 0 or 1 is too strict, and the model will be very easy to trigger the early stop.

Reply:
LGTM 😆!